app/commands/profiles/update.rb

```diff
@@ -0,0 +1,52 @@
+module Profiles
```

Comment:
So tbh I am not a big fan of this type of abstraction for the most part. To solve one of my problems, the `call!` should either return nothing or raise an error. Alternatively, you could make it `call` which can return true/false, but I much prefer raising custom errors specific to what went wrong.

(Another solution would be to pass in the controller in via self).

Now, given that we will be having a resume upload needed in the `user_mentee_application` I think tailoring to that specifically, using a service object is acceptable. That being said, what should be passed in here is most likely a `user` and a `resume`.

There might be some complications with figuring out how to properly redirect, but we can figure that out later.

Reply:
When you have some time, I'd be curious to hear a bit more about the time using this type of command/service object abstraction to encapsulate domain logic ended up causing an issue for you.

I'm also unclear about why you prefer `call!` to return nothing or raising an error? I can understand wanting to bubble up an error, though as long as that error gets handled and reported (we probably could look into rollbar, honeybadger, or something similar), seems like gracefully recovering from it is important.

I have historically referenced the [stitchfix post on anatomy of a service object](https://multithreaded.stitchfix.com/blog/2015/06/02/anatomy-of-service-objects-in-rails/) as rough guidelines and specifically like returning rich objects. In this case, I chose to use symbols over a more complex object to simplify the implementation so I could solicit feedback more quickly. Additionally, it does mirror other libraries that follow similar patterns, like [rectify](https://github.com/andypike/rectify).  However, one thing I like about the symbols is that they they provide more context about what the method did rather than silently passing, raising an error, or true/false.

I actually do like failing noisily, so I don't mind rescuing and reraising if the Error is not what I expect it to be, but am wary of it until it's clear that we have a solid error reporting system in place that doesn't completely disrupt service to the end users.

Reply:
Whenever i've done service objects I generally follow the `call!` pattern (cause I like to fail noisily). (There are times when i'd have different methods if different places need different data from the same source).

My thought process goes like this, the `!` means it will raise some error on failure and it is your job as the caller to catch it and handle appropriately. If it returns nothing, that means the service succeeded in what it was doing (this can also mean the service didn't need to do anything because making them idempotent is nice). 

As for reporting, yeah we should prob implement that at some point. I generally tend to use Sentry and have had good results. 

Sidenote:

Now we could include Dry Structs/ Dry types or ways to returned typed data to keep things consistent but I still don't see much value adding more levels of complexity.